### PR TITLE
RBAC: Add permissions to install and configure plugins

### PR DESCRIPTION
**Pull request by gamab**
**What this PR does / why we need it**:
This PR adds RBAC permissions check to install/uninstall plugins (`plugins:install`) and enable/disable plugins (`plugins:write`).

*Note on the Legacy accesscontrol*:
* Can see the Configuration > Plugins page: Org Admins and Grafana Admins if `cfg.PluginAdminEnabled`
* Can install plugins: Grafana Admins if `cfg.PluginAdminEnabled && !cfg.PluginAdminExternalManageEnabled`
* Can configure (toggle, edit settings) plugins: `Org Admins`
* Can list non-core plugins: `Grafana Admins` and `Org Admins`

**Which issue(s) this PR fixes**:

Partially fixes  #https://github.com/grafana/grafana-enterprise/issues/3494

**Special notes for your reviewer**:

Here is a quick demo of how assigning the new plugin roles works:

demo

### CODE REVIEW

File: pkg/plugins/guardian/guardian.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
+package guardian
```
Comment (gamab) [inline pkg/plugins/guardian/guardian.go:1]: Note that this guardian will be removed or largely simplefied once we introduce `plugins:read` and `plugins.settings:read` as knowing if a user can view a plugin and its settings will be a direct check and not derived from other permissions.

File: pkg/plugins/accesscontrol.go [hunk 2]
```diff
@@ -26,5 +38,35 @@ func DeclareRBACRoles(acService ac.AccessControl) error {
...
+			Group:       "Plugins",
+			Permissions: []ac.Permission{
+				{Action: ActionInstall},
+			},
+		},
+		Grants: []string{ac.RoleGrafanaAdmin},
+	}
+
+	// Remove possibility to install plugins if `plugin_admin_enable` is false
+	if !cfg.PluginAdminEnabled {
```
Comment (wbrowne) [inline pkg/plugins/accesscontrol.go:67]: If I'm not mistaken, we'll also need to verify if `|| cfg.PluginAdminExternalManageEnabled` here - a configuration we use in Cloud that brings users to grafana.com to manage their plugins instead

File: pkg/plugins/accesscontrol.go [hunk 3]
```diff
@@ -3,17 +3,29 @@ package plugins
 import (
 	"github.com/grafana/grafana/pkg/models"
 	ac "github.com/grafana/grafana/pkg/services/accesscontrol"
+	"github.com/grafana/grafana/pkg/setting"
 )
 
 const (
+	// Plugins actions
+	ActionInstall       = "plugins:install"
+	ActionSettingsWrite = "plugins.settings:write"
```
Comment (kalleep) [inline pkg/plugins/accesscontrol.go:12]: Maybe we should rename this one to `plugins:write`. That action would allow users to update settings and enable / disable plugin

File: pkg/api/index.go [hunk 4]
```diff
@@ -378,7 +378,11 @@ func (hs *HTTPServer) setupConfigNodes(c *models.ReqContext) ([]*dtos.NavLink, e
 		})
 	}
 
-	if c.OrgRole == org.RoleAdmin || (hs.Cfg.PluginAdminEnabled && ac.ReqGrafanaAdmin(c)) {
+	pluginAccess := func(rc *models.ReqContext) bool {
+		return rc.OrgRole == org.RoleAdmin ||
+			hs.Cfg.PluginAdminEnabled && !hs.Cfg.PluginAdminExternalManageEnabled && rc.IsGrafanaAdmin
```
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/api/index.go:383]: Do I understand correctly that this is guarding the entire plugins UI section @gamab? `hs.Cfg.PluginAdminExternalManageEnabled` is only used when it comes to installing/uninstalling

[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/api/index.go]: >If we say that when hs.Cfg.PluginAdminExternalManageEnabled we prevent installing/uninstalling plugins from within Grafana, should server admin still access the plugins UI?

Yes they should be able to. `PluginAdminExternalManageEnabled` is only for controlling what the install/uninstall buttons do - so they should still be able to see the content of the plugins catalog and each plugin details page

**Review by Jguer — APPROVED** [review]

**Comment by gamab** [comment]
Hey @wbrowne, I'd like to merge this one tomorrow, do you see any issue with it?

Note: The work we are doing in https://github.com/grafana/grafana/pull/54845/ should simplify lots of the RBAC checks written in this one :)

**Review by wbrowne — APPROVED** [review]
LGTM

**Comment by bhks** [comment]
Hi @gamab ,

Please correct me if I am wrong. As per my understanding the  `IsGrafanaAdmin` have the highest permissions. and does not require the && with the config `PluginAdminEnabled` right ?

Based on the operator [precedence](https://www.tutorialspoint.com/go/go_operators_precedence.htm) `&&` will be processed first before `||` and hence this code change might cause issues in auth behavior.

My thinking as per my limited knowledge is that OrgAdmin require the binding with the flag/config ` `PluginAdminEnabled` whether  `GrafanaAdmin` Have enabled for Org Admin to perform Plugin operations.

```
func ReqCanAdminPlugins(cfg *setting.Cfg) func(rc *models.ReqContext) bool {
	// Legacy handler that protects access to the Configuration > Plugins page
	return func(rc *models.ReqContext) bool {
		return rc.OrgRole == org.RoleAdmin || cfg.PluginAdminEnabled && rc.IsGrafanaAdmin
	}
}
```

**Comment by mgyongyosi** [comment]
Hi @bhks,

Thank you for your question and the detailed description! 

> Please correct me if I am wrong. As per my understanding the IsGrafanaAdmin have the highest permissions. and does not require the && with the config PluginAdminEnabled right?

IsGrafanaAdmin has the highest permission, you are correct. However, `PluginAdminEnabled` is a config parameter that specifies whether plugins can be installed or not and it is not related to the user permissions.

We reviewed our implementation again and we think it's working correctly: The `ReqCanAdminPlugins` function is used to check if the /plugins page is accessible to the current user and we have one UI for both installing/uninstalling/updating and configuring plugins. If you check the documentation of the [Plugin Catalog](https://grafana.com/docs/grafana/latest/administration/plugin-management/#plugin-catalog) you can see that Org Admins "Can configure app plugins" so the function should return true (and the link should be visible in the navigation) when the current user is an Org Admin regardless of the `PluginAdminEnabled` settings, because Org Admins can configure the plugins too. Furthermore, the function should return true (and make the link visible) when the current user is a Grafana Admin and the `PluginAdminEnabled` config parameter is set to true, because GrafanaAdmins can only install/uninstall/update plugins, so it wouldn't make sense to show the link when `PluginAdminEnabled` is turned off.

I hope the previous clarifies how the check works!

**Comment by bhks** [comment]
Thanks @mgyongyosi for your reply. I didn't notice I had a response on this issue.

### Few clarifying questions :

1. Does value of `PluginAdminEnabled ` matter for ServerAdmin ?
2. When `PluginAdminEnabled` is `false` Org Admin can't see Install/Uninstall button ?
3. Does value of `PluginAdminEnabled` also affect plugin configuration updates by OrgAdmin or just installation operation ?

**Comment by gamab** [comment]
Hey @bhks, my apologies for the delay :pray: I took a look at your questions and at @mgyongyosi response and I second his statements :slightly_smiling_face:. 

**TL;DR;** the [`ReqCanAdminPlugins`](https://github.com/grafana/grafana/blob/main/pkg/services/pluginsintegration/pluginaccesscontrol/accesscontrol.go#L25-L30) check is used to know if we can display the admin view for plugins.
By default, organization admins [can access](https://github.com/grafana/grafana/blob/main/pkg/services/pluginsintegration/pluginaccesscontrol/accesscontrol.go#L45-L56) this view to configure/enable/disable plugins. Only server admins [are allowed to install plugins](https://github.com/grafana/grafana/blob/main/pkg/services/pluginsintegration/pluginaccesscontrol/accesscontrol.go#L57-L68). When `PluginAdminEnabled` is false, then [no one can install plugins](https://github.com/grafana/grafana/blob/main/pkg/services/pluginsintegration/pluginaccesscontrol/accesscontrol.go#L70-L72), only configuring/enabling/disabling installed plugins is available.

----

As for your questions, I'll try to answer as best as I can, feel free to tell me if something still remains unclear :slightly_smiling_face: :

> ##Few clarifying questions :##
> Does value of PluginAdminEnabled  matter for ServerAdmin ?

Yes it does, if `PluginAdminEnabled` is true, Server Admins can click the install/uninstall/update buttons:
image
With `PluginAdminDisabled` false, Server Admins cannot install/uninstall/update plugins.
image

> When PluginAdminEnabled is false Org Admin can't see Install/Uninstall button ?

Installing a plugin is a [server admin right](https://github.com/grafana/grafana/blob/main/pkg/services/pluginsintegration/pluginaccesscontrol/accesscontrol.go#L57-L72), an OrgAdmin that is not a server admin is not allowed to install/uninstall/update plugins. Therefore, Org Admin will never see the Install/Uninstall button, no matter the value of `PluginAdminEnabled`.
image

> Does value of PluginAdminEnabled also affect plugin configuration updates by OrgAdmin or just installation operation ?

`PluginAdminEnabled` has no affect on plugin configuration updates. Independently of its value, OrgAdmins will be able to configure installed plugins.
image

----



*Also note that this PR did not change the check in question `rc.OrgRole == org.RoleAdmin || (cfg.PluginAdminEnabled && rc.IsGrafanaAdmin)`.*

**Comment by bhks** [comment]
Hi @gamab, 

You both are fantastic and awesome!! Thanks you taking the time to explain this in a way a 5 year old can understand.
